src/main/java/net/pms/dlna/DLNAThumbnailInputStream.java

```diff
@@ -0,0 +1,348 @@
+	 * @return The populated {@link DLNAThumbnailInputStream} or {@code null}
+	 *         if the source image could not be parsed.
+	 * @throws IOException if the operation fails.
+	 */
+	public static DLNAThumbnailInputStream toThumbnailInputStream(
+		InputStream inputStream,
+		int width,
+		int height,
+		ScaleType scaleType,
+		ImageFormat outputFormat,
+		boolean padToSize
+	) throws IOException {
+		DLNAThumbnail thumbnail = DLNAThumbnail.toThumbnail(inputStream, width, height, scaleType, outputFormat, padToSize);
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+    /**
+     * Creates a {@link DLNAThumbnailInputStream} where it uses
+     * {@code thumbnail}'s buffer as its buffer array. The buffer array is
+     * not copied.
+     *
+     * @param thumbnail the input thumbnail.
+	 * @return The populated {@link DLNAThumbnailInputStream} or {@code null}
+	 *         if the source thumbnail is {@code null}.
+     */
+
+	public static DLNAThumbnailInputStream toThumbnailInputStream(DLNAThumbnail thumbnail) {
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+	/**
+	 * Don't call this from outside this class or subclass, use
+	 * {@link DLNAThumbnailInputStream#toThumbnailInputStream(DLNAThumbnail)}
+	 * which handles {@code null} input.
+	 *
+	 * @param thumbnail the input thumbnail
+	 *
+	 * @throws NullPointerException if {@code thumbnail} is {@code null}.
+	 */
+    protected DLNAThumbnailInputStream(DLNAThumbnail thumbnail) {
+    	super(thumbnail.getBytes(false));
+    	this.imageInfo = thumbnail.getImageInfo();
+    	this.profile = thumbnail.getDLNAImageProfile();
+    }
+
+	/**
+	 * Converts and scales a thumbnail according to the given
+	 * {@link DLNAImageProfile}. Preserves aspect ratio. Format support is
+	 * limited to that of {@link ImageIO}.
+	 *
+	 * @param outputProfile the DLNA media profile to adhere to for the output.
+	 * @param updateMetadata whether or not new metadata should be updated
+	 *                       after image transformation. This should only be
+	 *                       disabled if the output image won't be kept/reused.
+	 * @param padToSize Whether padding should be used if source aspect doesn't
+	 *                  match target aspect.
+	 * @return The scaled and/or converted thumbnail, {@code null} if the
+	 *         source is {@code null}.
+	 * @exception IOException if the operation fails.
+	 */
+	public DLNAThumbnailInputStream transcode(
+		DLNAImageProfile outputProfile,
+		boolean updateMetadata,
+		boolean padToSize
+	) throws IOException {
+		DLNAThumbnail thumbnail;
+		thumbnail = (DLNAThumbnail) ImagesUtil.transcodeImage(
+			this.getBytes(false),
+			outputProfile,
+			updateMetadata,
+			true,
+			padToSize);
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+	/**
+	 * @return The bytes of this thumbnail.
+	 */
+	@SuppressFBWarnings("EI_EXPOSE_REP")
```

Comment:
@Nadahar Thank you very much for this great explanation :smile: 

I was not aware of this, so after reading you i searched and discovered that only [ArrayList](http://www.baeldung.com/java-immutable-list) was natively able to be immutable.
So, if i believe [this guy](http://www.xinotes.net/notes/note/1385/), your work is great :+1: